File: chapter 05/recipe 5.5/ComVariantsVB/ComVariantsTest.vb
Imports DniDataTypesLib
Imports System.Runtime.InteropServices

Module ComVariantsTest

    Sub Main()

        'create an instance of the COM object				
        Dim comObj As DniDataTypesObj = New DniDataTypesObj()
        Dim desc As String = String.Empty

        'a variable used to return any VARIANT value
        Dim result As Object = Nothing

        Dim testVariantBool As Boolean = True
        desc = comObj.UseVariant(testVariantBool, result)
        Console.WriteLine("Test Variant Bool: {0}, {1}, {2}", _
            testVariantBool.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantInt As Integer = 123
        desc = comObj.UseVariant(testVariantInt, result)
        Console.WriteLine("Test Variant Int: {0}, {1}, {2}", _
            testVariantInt.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantLong As Long = 123
        desc = comObj.UseVariant(testVariantLong, result)
        Console.WriteLine("Test Variant Long: {0}, {1}, {2}", _
            testVariantLong.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantString As String = "test string"
        desc = comObj.UseVariant(testVariantString, result)
        Console.WriteLine("Test Variant String: {0}, {1}, {2}", _
            testVariantString.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantObject As Object = New Object()
        desc = comObj.UseVariant(testVariantObject, result)
        Console.WriteLine("Test Variant Object: {0}, {1}, {2}", _
            testVariantObject.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantNull As Object = Nothing
        desc = comObj.UseVariant(testVariantNull, result)
        Dim returnType As String
        If result = Nothing Then
            returnType = "null"
        Else
            returnType = result.GetType().Name
        End If
        Console.WriteLine("Test Variant Null: {0}, {1}, {2}", _
            "null", returnType, desc)

        Dim testVariantDouble As Double = 123.45
        desc = comObj.UseVariant(testVariantDouble, result)
        Console.WriteLine("Test Variant Double: {0}, {1}, {2}", _
            testVariantDouble.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantDecimal As Decimal = 123.45D
        desc = comObj.UseVariant(testVariantDecimal, result)
        Console.WriteLine("Test Variant Decimal: {0}, {1}, {2}", _
            testVariantDecimal.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantDate As DateTime = New DateTime(2005, 12, 31)
        desc = comObj.UseVariant(testVariantDate, result)
        Console.WriteLine("Test Variant Date: {0}, {1}, {2}", _
            testVariantDate.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantCurrency As CurrencyWrapper _
            = New CurrencyWrapper(123.45D)
        desc = comObj.UseVariant(testVariantCurrency, result)
        Console.WriteLine("Test Variant Currency: {0}, {1}, {2}", _
            testVariantCurrency.GetType().Name, _
            result.GetType().Name, desc)

        Dim testVariantIntPtr As IntPtr = IntPtr.Zero
        desc = comObj.UseVariant(testVariantIntPtr, result)
        Console.WriteLine("Test Variant IntPtr: {0}, {1}, {2}", _
            testVariantIntPtr.GetType().Name, _
            result.GetType().Name, desc)

        'wait for input
        Console.WriteLine("Press any key to exit")
        Console.Read()

    End Sub

End Module

File: chapter 05/recipe 5.5/ComVariants/ComVariantsTest.cs
using System;
using System.Runtime.InteropServices;

using DniDataTypesLib;

namespace ComVariants
{
	class ComVariantsTest
	{
		static void Main(string[] args)
		{
			//create an instance of the COM object				
			DniDataTypesObj comObj = new DniDataTypesObj();
			string desc = string.Empty;

			//a variable used to return any VARIANT value
			Object result = null;

			Boolean testVariantBool	= true;				
			desc = comObj.UseVariant(testVariantBool, out result);
			Console.WriteLine("Test Variant Bool: {0}, {1}, {2}",
				testVariantBool.GetType().Name, 
				result.GetType().Name, desc);

			int testVariantInt = 123;
			desc = comObj.UseVariant(testVariantInt, out result);
			Console.WriteLine("Test Variant Int: {0}, {1}, {2}",
				testVariantInt.GetType().Name,
				result.GetType().Name, desc);

			long testVariantLong = 123;
			desc = comObj.UseVariant(testVariantLong, out result);
			Console.WriteLine("Test Variant Long: {0}, {1}, {2}",
				testVariantLong.GetType().Name,
				result.GetType().Name, desc);

			string testVariantString = "test string";
			desc = comObj.UseVariant(testVariantString, out result);
			Console.WriteLine("Test Variant String: {0}, {1}, {2}",
				testVariantString.GetType().Name,
				result.GetType().Name, desc);

			object testVariantObject = new object();
			desc = comObj.UseVariant(testVariantObject, out result);
			Console.WriteLine("Test Variant Object: {0}, {1}, {2}",
				testVariantObject.GetType().Name,
				result.GetType().Name, desc);

			object testVariantNull = null;
			desc = comObj.UseVariant(testVariantNull, out result);
			string returnType;
			if (result == null)
			{
				returnType = "null";
			}
			else
			{
				returnType = result.GetType().Name;
			}
			Console.WriteLine("Test Variant Null: {0}, {1}, {2}",
				"null", returnType, desc); 

			double testVariantDouble = 123.45;
			desc = comObj.UseVariant(testVariantDouble, out result);
			Console.WriteLine("Test Variant Double: {0}, {1}, {2}",
				testVariantDouble.GetType().Name,
				result.GetType().Name, desc);

			decimal testVariantDecimal = 123.45m;
			desc = comObj.UseVariant(testVariantDecimal, out result);
			Console.WriteLine("Test Variant Decimal: {0}, {1}, {2}",
				testVariantDecimal.GetType().Name,
				result.GetType().Name, desc);

			DateTime testVariantDate = new DateTime(2005, 12, 31);
			desc = comObj.UseVariant(testVariantDate, out result);
			Console.WriteLine("Test Variant Date: {0}, {1}, {2}",
				testVariantDate.GetType().Name,
				result.GetType().Name, desc);

			CurrencyWrapper testVariantCurrency = new CurrencyWrapper(123.45m);
			desc = comObj.UseVariant(testVariantCurrency, out result);
			Console.WriteLine("Test Variant Currency: {0}, {1}, {2}",
				testVariantCurrency.GetType().Name,
				result.GetType().Name, desc);

			IntPtr testVariantIntPtr = IntPtr.Zero;
			desc = comObj.UseVariant(testVariantIntPtr, out result);
			Console.WriteLine("Test Variant IntPtr: {0}, {1}, {2}",
				testVariantIntPtr.GetType().Name,
				result.GetType().Name, desc);

			//wait for input
			Console.WriteLine("Press any key to exit");
			Console.Read(); 	    

		}
	}
}
